buku

```diff
@@ -1492,132 +1545,131 @@ class BukuDb:
+            order: List[str] = ['+id'],
     ) -> List[BookmarkVar]:
         """Search DB for entries where tags, URL, or title fields match keywords.
 
         Parameters
         ----------
         keywords : list of str
             Keywords to search.
+        order : list of str
+            Order description (fields from JSON export or DB, prepended with '+'/'-' for ASC/DESC).
+            Note: this applies to fields with the same number of matched keywords.
         all_keywords : bool
-            True to return records matching ALL keywords.
             False (default value) to return records matching ANY keyword.
+            True to return records matching ALL keywords. This also enables special
+            behaviour when keywords in (['blank'], ['immutable']).
         deep : bool
             True to search for matching substrings. Default is False.
+        markers : bool
+            True to use prefix markers for different fields. Default is False.
         regex : bool
             Match a regular expression if True. Default is False.
+            Overrides deep, all_keywords, and comma matching in tags with markers.
 
         Returns
         -------
         list
             List of search results.
         """
-        if not keywords:
+        _order = self._order(order)
+        clauses, qargs = [], []
+        for keyword in keywords:
+            tokens = self._search_tokens(keyword, deep=deep, markers=markers)
+            clause, args = self._search_clause(tokens, regex=regex)
+            if clause and args:
+                clauses += [f'({clause})']
+                qargs += args
+        if not qargs:
             return []
 
-        # Deep query string
-        q1 = ("(tags LIKE ('%' || ? || '%') OR "
-              "URL LIKE ('%' || ? || '%') OR "
-              "metadata LIKE ('%' || ? || '%') OR "
-              "desc LIKE ('%' || ? || '%')) ")
-        # Non-deep query string
-        q2 = ('(tags REGEXP ? OR '
-              'URL REGEXP ? OR '
-              'metadata REGEXP ? OR '
-              'desc REGEXP ?) ')
-        qargs = []  # type: List[Any]
-
-        case_statement = lambda x: 'CASE WHEN ' + x + ' THEN 1 ELSE 0 END'
+        _count = lambda x: f'CASE WHEN {x} THEN 1 ELSE 0 END'
         if regex:
-            q0 = 'SELECT id, url, metadata, tags, desc, flags FROM (SELECT *, '
-            for token in keywords:
-                if not token:
-                    continue
-
-                q0 += case_statement(q2) + ' + '
-                qargs += (token, token, token, token,)
-
-            if not qargs:
-                return []
-
-            q0 = q0[:-3] + ' AS score FROM bookmarks WHERE score > 0 ORDER BY score DESC)'
+            query = ('SELECT id, url, metadata, tags, desc, flags\nFROM (SELECT *, (' +
+                     '\n    + '.join(map(_count, clauses)) +
+                     f') AS score\n  FROM bookmarks WHERE score > 0 ORDER BY score DESC, {_order})')
```

Comment:
The query-generating code itself is much more compact now.

(As for the `\n`s, they make no difference to the DB engine, but reading logged queries is easier when they're formatted.)